components/script/dom/htmlmetaelement.rs

```diff
@@ -81,6 +111,97 @@ impl HTMLMetaElement {
             }
         }
     }
+
+    /// <https://html.spec.whatwg.org/multipage/#shared-declarative-refresh-steps>
+    fn declarative_refresh(&self) {
+        // 2
+        let content = self.Content();
+        // 1
+        if !content.is_empty() {
+            // 3
+            self.shared_declarative_refresh_steps(content);
+        }
+    }
+
+    /// <https://html.spec.whatwg.org/multipage/#shared-declarative-refresh-steps>
+    fn shared_declarative_refresh_steps(&self, content: DOMString) {
+        // 1
+        let document = document_from_node(self);
+        if document.will_declaratively_refresh() {
+            return;
+        }
+
+        // 2-11
+        lazy_static::lazy_static! {
+            static ref REFRESH_REGEX: Regex = Regex::new(
+                r#"(?x)
+                ^
+                \s* # 3
+                ((?<time>\d+)\.?|\.) # 5-6
+                [0-9.]* # 8
+                (
+                    (;|,| ) # 10.1
+                    \s* # 10.2
+                    (;|,)? # 10.3
+                    \s* # 10.4
+                    (
+                        (U|u)(R|r)(L|l) # 11.2-11.4
+                        \s*=\s* # 11.5-11.7
+                        ('(?<url1>.*?)'?|"(?<url2>.*?)"?|(?<url3>[^'"].*)) # 11.8 - 11.10
+                        |
+                        (?<url4>.*)
+                    )?
+                )?
+                $
+            "#,
+            )
+            .unwrap();
+        }
+        let mut url_record = document.url();
+        let captures = if let Some(captures) = REFRESH_REGEX.captures(content.as_bytes()) {
+            captures
+        } else {
+            return;
+        };
+        let time = if let Some(time_string) = captures.name("time") {
+            u64::from_str(&String::from_utf8_lossy(time_string.as_bytes())).unwrap_or(0)
+        } else {
+            0
+        };
+        let captured_url = captures.name("url1").or_else(|| {
+            captures
+                .name("url2")
+                .or_else(|| captures.name("url3").or_else(|| captures.name("url4")))
+        });
+        if let Some(url_match) = captured_url {
+            url_record = if let Ok(url) = ServoUrl::parse_with_base(
+                Some(&url_record),
+                &String::from_utf8_lossy(url_match.as_bytes()),
+            ) {
+                url
+            } else {
+                return;
+            }
+        }
+        // 12-13
+        if document.completely_loaded() {
+            // TODO: handle active sandboxing flag
+            let window = window_from_node(self);
+            window.upcast::<GlobalScope>().schedule_callback(
+                OneshotTimerCallback::RefreshRedirectDue(RefreshRedirectDue {
+                    window: window.clone(),
+                    url: DomRefCell::new(url_record),
+                }),
+                MsDuration::new(dbg!(time.saturating_mul(1000))),
```

Comment:
Is this dbg still necessary?

Reply:
I accidentally left that in from debugging, removed.